File: src/utils/crypto.ts
import crypto from 'crypto';

const ALGORITHM = 'aes-256-gcm';
const PBKDF2_ITERATIONS = 600_000;
const SALT_LENGTH = 32;
const IV_LENGTH = 12;
const AUTH_TAG_LENGTH = 16;
const KEY_LENGTH = 32;
const PREFIX = 'encrypted::v1:';

/**
 * Check whether a string value is an encrypted CodexCLI value.
 */
export function isEncrypted(value: string): boolean {
  return typeof value === 'string' && value.startsWith(PREFIX);
}

/**
 * Encrypt a plaintext string using AES-256-GCM with a password-derived key.
 * Returns a prefixed base64 string: `encrypted::v1:<base64(salt+iv+authTag+ciphertext)>`
 */
export function encryptValue(plaintext: string, password: string): string {
  const salt = crypto.randomBytes(SALT_LENGTH);
  const iv = crypto.randomBytes(IV_LENGTH);
  const key = crypto.pbkdf2Sync(password, salt, PBKDF2_ITERATIONS, KEY_LENGTH, 'sha256');

  const cipher = crypto.createCipheriv(ALGORITHM, key, iv);
  const encrypted = Buffer.concat([cipher.update(plaintext, 'utf8'), cipher.final()]);
  const authTag = cipher.getAuthTag();

  const payload = Buffer.concat([salt, iv, authTag, encrypted]);
  return PREFIX + payload.toString('base64');
}

/**
 * Decrypt an encrypted CodexCLI value using a password.
 * Throws on wrong password or corrupted data.
 */
export function decryptValue(encrypted: string, password: string): string {
  if (!isEncrypted(encrypted)) {
    throw new Error('Value is not encrypted.');
  }

  const payload = Buffer.from(encrypted.slice(PREFIX.length), 'base64');

  const minLength = SALT_LENGTH + IV_LENGTH + AUTH_TAG_LENGTH;
  if (payload.length < minLength) {
    throw new Error('Corrupted encrypted data.');
  }

  const salt = payload.subarray(0, SALT_LENGTH);
  const iv = payload.subarray(SALT_LENGTH, SALT_LENGTH + IV_LENGTH);
  const authTag = payload.subarray(SALT_LENGTH + IV_LENGTH, SALT_LENGTH + IV_LENGTH + AUTH_TAG_LENGTH);
  const ciphertext = payload.subarray(SALT_LENGTH + IV_LENGTH + AUTH_TAG_LENGTH);

  const key = crypto.pbkdf2Sync(password, salt, PBKDF2_ITERATIONS, KEY_LENGTH, 'sha256');

  const decipher = crypto.createDecipheriv(ALGORITHM, key, iv);
  decipher.setAuthTag(authTag);

  try {
    const decrypted = Buffer.concat([decipher.update(ciphertext), decipher.final()]);
    return decrypted.toString('utf8');
  } catch {
    throw new Error('Decryption failed. Wrong password or corrupted data.');
  }
}

/**
 * Replace encrypted leaf values with '[encrypted]' for safe display/export.
 */
export function maskEncryptedValues(data: Record<string, unknown>): Record<string, unknown> {
  const result: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(data)) {
    if (typeof value === 'string' && isEncrypted(value)) {
      result[key] = '[encrypted]';
    } else if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
      result[key] = maskEncryptedValues(value as Record<string, unknown>);
    } else {
      result[key] = value;
    }
  }
  return result;
}

File: codexcli-go/internal/crypto/crypto.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"strings"

	"golang.org/x/crypto/pbkdf2"
)

const (
	prefix          = "encrypted::v1:"
	saltLength      = 32
	ivLength        = 12
	authTagLength   = 16
	keyLength       = 32
	pbkdf2Iterations = 600_000
	minPayloadLen   = saltLength + ivLength + authTagLength
)

// IsEncrypted returns true if the value has the encryption prefix.
func IsEncrypted(value string) bool {
	return strings.HasPrefix(value, prefix)
}

// EncryptValue encrypts plaintext using AES-256-GCM with PBKDF2 key derivation.
// The output format is: encrypted::v1:<base64(salt + iv + authTag + ciphertext)>
func EncryptValue(plaintext, password string) (string, error) {
	salt := make([]byte, saltLength)
	if _, err := rand.Read(salt); err != nil {
		return "", err
	}

	iv := make([]byte, ivLength)
	if _, err := rand.Read(iv); err != nil {
		return "", err
	}

	key := pbkdf2.Key([]byte(password), salt, pbkdf2Iterations, keyLength, sha256.New)

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}

	aesGCM, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	// Seal appends ciphertext + authTag
	sealed := aesGCM.Seal(nil, iv, []byte(plaintext), nil)

	// sealed = ciphertext + authTag (GCM appends tag at end)
	// We need: salt + iv + authTag + ciphertext (to match TypeScript format)
	ciphertextLen := len(sealed) - authTagLength
	ciphertext := sealed[:ciphertextLen]
	authTag := sealed[ciphertextLen:]

	payload := make([]byte, 0, saltLength+ivLength+authTagLength+ciphertextLen)
	payload = append(payload, salt...)
	payload = append(payload, iv...)
	payload = append(payload, authTag...)
	payload = append(payload, ciphertext...)

	return prefix + base64.StdEncoding.EncodeToString(payload), nil
}

// DecryptValue decrypts an encrypted value string.
func DecryptValue(encrypted, password string) (string, error) {
	if !IsEncrypted(encrypted) {
		return "", errors.New("value is not encrypted")
	}

	payload, err := base64.StdEncoding.DecodeString(encrypted[len(prefix):])
	if err != nil {
		return "", errors.New("corrupted encrypted data")
	}

	if len(payload) < minPayloadLen {
		return "", errors.New("corrupted encrypted data")
	}

	salt := payload[:saltLength]
	iv := payload[saltLength : saltLength+ivLength]
	authTag := payload[saltLength+ivLength : saltLength+ivLength+authTagLength]
	ciphertext := payload[saltLength+ivLength+authTagLength:]

	key := pbkdf2.Key([]byte(password), salt, pbkdf2Iterations, keyLength, sha256.New)

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", errors.New("decryption failed. Wrong password or corrupted data")
	}

	aesGCM, err := cipher.NewGCM(block)
	if err != nil {
		return "", errors.New("decryption failed. Wrong password or corrupted data")
	}

	// Reassemble: ciphertext + authTag (Go's GCM expects tag appended)
	sealed := make([]byte, 0, len(ciphertext)+authTagLength)
	sealed = append(sealed, ciphertext...)
	sealed = append(sealed, authTag...)

	plaintext, err := aesGCM.Open(nil, iv, sealed, nil)
	if err != nil {
		return "", errors.New("decryption failed. Wrong password or corrupted data")
	}

	return string(plaintext), nil
}

// MaskEncryptedValues replaces encrypted values with "[encrypted]" in a map.
func MaskEncryptedValues(data map[string]any) map[string]any {
	result := make(map[string]any, len(data))
	for k, v := range data {
		switch val := v.(type) {
		case string:
			if IsEncrypted(val) {
				result[k] = "[encrypted]"
			} else {
				result[k] = val
			}
		case map[string]any:
			result[k] = MaskEncryptedValues(val)
		default:
			result[k] = v
		}
	}
	return result
}
